crates/server/src/store/file/authority.rs

```diff
@@ -54,6 +58,66 @@ impl FileAuthority {
         InMemoryAuthority::new(origin, records, zone_type, allow_axfr).map(Self)
     }
 
+    /// Read given file line by line and recursively invokes reader for
+    /// $INCLUDE directives
+    ///
+    /// TODO: it looks hacky as far we effectively duplicate parser's functionallity
+    /// (at least partially) and performing lexing twice.
+    /// Better solution requires us to change lexer to deal
+    /// with Lines-like iterator instead of String buf (or capability to combine a few
+    /// lexer instances into a single lexer).
+    ///
+    /// TODO: we should probably track depth of recursion here and set max level (or
```

Comment:
Agree on the max recursion concern, though the outcome would be the same, stack-overflow, panic. We'd just panic with a better error message, but definitely should do it. This would also handle any unintended loops...

Reply:
Yep, loop is probably the easiest way to get recursion limit. I will add 2 different errors. One for too deep nesting and another for loops.